SwiftyJSONAPI/JSONAPIResource.swift

```diff
@@ -102,13 +105,40 @@ public class JSONAPIResource: JSONPrinter {
                 
                 resource.attributes    = includedResource.attributes
                 resource.relationships = includedResource.relationships
+            
                 
-                if !resource.relationships.isEmpty {
+                if !resource.relationships.isEmpty, resource.shouldLoadResource(withCachedResources: &cachedResources) {
+                    
                     resource.parent = self.parent
-                    resource.loadResources(withIncludedResources: includedResources)
+                    resource.loadResources(withIncludedResources: includedResources, cachedResources: &cachedResources)
                 }
+                
                 resource.loaded = .Loaded
             }
         }
     }
 }
+
+private extension JSONAPIResource {
+    
+    // Checks if a resource has been loaded already, prevents bidirectional relationships from being recursively called
+    func shouldLoadResource(withCachedResources cachedResources: inout CachedResources) -> Bool {
+        
+        if var cachedResourceIds = cachedResources[type] {
+            
+            if cachedResourceIds.contains(id) {
+                
+                return false
+            } else {
+                
+                cachedResourceIds.insert(id)
+                cachedResources[type] = cachedResourceIds
+                return true
+            }
+        } else {
+            
+            cachedResources[type] = Set([id])
+            return true
```

Comment:
Did you forget to insert the id here as well to the new set?

Reply:
In this case it means that this type was not present in the dictionary, so we create a new key value.

You can see the check for this type here: https://github.com/thomassnielsen/SwiftyJSONAPI/pull/13/files#diff-17206e60ee1273002aceab1589875e69R127

Reply:
oh didn't notice the `[id]` 👍